pkg/parse/source.go

```diff
@@ -96,10 +96,39 @@ func (o *files) readConfigFiles(state *sourceState) status.Error {
 	if err != nil {
 		return status.PathWrapError(errors.Wrap(err, "listing files in the configs directory"), syncDir.OSPath())
 	}
+
+	commitChanged, newCommit := o.checkCommitChanged(p, state)
+	if commitChanged && newCommit != "" {
+		return status.InternalError(fmt.Sprintf("source commit changed while listing files, was %s, now %s. It will be retried in the next sync", state.commit, newCommit))
+	}
+
 	state.files = fileList
 	return nil
 }
 
+func (o *files) checkCommitChanged(p Parser, state *sourceState) (bool, string) {
+	// compute the source commit and compare with the given sourceCommit
+	opts := p.options()
+	absHydratedRoot, err := cmpath.AbsoluteOS(opts.HydratedRoot)
+	if err != nil {
+		klog.Errorf("not able to get absolute HydratedRoot while checking commit changed after listing files: %v", err)
+		return false, ""
```

Comment:
Same here. If it failed to get the absolute path, it might proceed with the new commit.

Reply:
Added error in return and check